source/grid/grid_tools.cc

```diff
@@ -1637,6 +1635,187 @@ namespace GridTools
+
+    // compute cell radius for each boundary cell of the predicate subdomain
+    for ( typename std::vector<typename MeshType::active_cell_iterator>::const_iterator
+          boundary_cell_iterator  = subdomain_boundary_cells.begin();
+          boundary_cell_iterator != subdomain_boundary_cells.end(); ++boundary_cell_iterator )
+      {
+        boundary_cells_centers.push_back( (*boundary_cell_iterator)->center() );
+        boundary_cells_radii.push_back((*boundary_cell_iterator )->diameter() / 2.0);
+      }
+    AssertThrow( boundary_cells_radii.size() == boundary_cells_centers.size(),
+                 ExcInternalError());
+
+    // Find the cells that are within layer_thickness of predicate subdomain boundary
+    // distance but are inside the extended bounding box.
+    // Most cells might be outside the extended bounding box, so we could skip them.
+    // Those cells that are inside the extended bounding box but are not part of the
+    // predicate subdomain are possible candidates to be within the distance to the
+    // boundary cells of the predicate subdomain.
+    for ( typename MeshType::active_cell_iterator
+          cell = mesh.begin_active();
+          cell != mesh.end(); ++cell)
+      {
+        // Ignore all the cells that are in the predicate subdomain
+        if ( predicate(cell))
+          continue;
+
+        // define cell_radius as the farthest distance of cell vertices to cell center
+        const double cell_radius = cell->diameter() / 2.0;
+
+        bool cell_inside = true; // reset for each cell
+
+        // Faster to check to check with cell->center() instead of all of its vertices
+        for (unsigned int d = 0; d < spacedim; ++d)
+          cell_inside &= (cell->center()[d] + cell_radius > bounding_box.first[d])
+                         && (cell->center()[d] - cell_radius < bounding_box.second[d]);
+        // cell_inside is true if any of the cell vertices are inside the extended bounding box
+
+        // Ignore all the cells that are outside the extended bounding box
+        if ( cell_inside) // cell is
+          for ( unsigned int i =0; i< boundary_cells_radii.size(); ++i)
+            if ( (cell->center() - boundary_cells_centers[i]).norm_square()
+                 <  Utilities::fixed_power<2>( cell_radius + boundary_cells_radii[i] + layer_thickness) )
+              {
+                active_cell_layer_within_distance.push_back(cell);
+                break; // Exit the loop checking all the remaining subdomain boundary cells
+              }
+
+      }
+    return active_cell_layer_within_distance;
+  }
+
+
+  template <class MeshType>
+  std::vector<typename MeshType::active_cell_iterator>
+  compute_ghost_cell_layer_within_distance ( const MeshType &mesh, const double layer_thickness)
+  {
+    std_cxx11::function<bool (const typename MeshType::active_cell_iterator &)> predicate
+      = IteratorFilters::LocallyOwnedCell();
```

Comment:
This is a corner of C++ where I keep forgetting what happens. You are creating a temporary, unnamed variable on the rhs here. And then you create a reference to it when you bind it to the `predicate` variable. Does this reference go stale when the temporary variable gets destroyed? Does it get destroyed right away, or only when `predicate` gets destroyed? I can't remember, but it's probably best to avoid this situation.